python-checks/src/main/java/org/sonar/python/checks/PytzTimeZoneInDatetimeConstructorCheck.java

```diff
@@ -19,64 +19,65 @@
   }
 
-  private void checkContext(SubscriptionContext context) {
+  private void checkCallExpression(SubscriptionContext context) {
     CallExpression callExpression = (CallExpression) context.syntaxNode();
     Symbol calleeSymbol = callExpression.calleeSymbol();
 
     if (calleeSymbol != null && "datetime.datetime".equals(calleeSymbol.fullyQualifiedName())) {
-      ArgList argList = callExpression.argumentList();
-      if (argList != null) {
-        List<Argument> argumentList = argList.arguments();
-        argumentList.stream().filter(this::checkArgument)
-          .findAny().ifPresent(argument -> context.addIssue(argument, MESSAGE));
+      RegularArgument argument = TreeUtils.nthArgumentOrKeyword(-1, "tzinfo", callExpression.arguments());
+      if (argument == null) {
+        return;
       }
+      createIssue(argument, context);
     }
   }
 
-  private boolean checkArgument(Argument argument) {
-    if (argument.is(Tree.Kind.REGULAR_ARGUMENT)) {
-      RegularArgument regularArgument = ((RegularArgument) argument);
-      if (regularArgument.expression().is(Tree.Kind.CALL_EXPR)) {
-        CallExpression callExpression1 = (CallExpression) regularArgument.expression();
-        Symbol calleeSymbol = callExpression1.calleeSymbol();
-        return calleeSymbol != null && "pytz.timezone".equals(calleeSymbol.fullyQualifiedName());
-      } else if (regularArgument.expression().is(Tree.Kind.NAME)) {
-        return reachingDefinitionsAnalysis.valuesAtLocation((Name) regularArgument.expression()).stream()
-          .filter(expression -> expression.is(Tree.Kind.CALL_EXPR))
-          .map(CallExpression.class::cast)
-          .map(CallExpression::calleeSymbol)
-          .filter(Objects::nonNull)
-          .anyMatch(symbol -> "pytz.timezone".equals(symbol.fullyQualifiedName()));
+  private void createIssue(RegularArgument argument, SubscriptionContext context) {
+    if (argument.expression().is(Tree.Kind.CALL_EXPR)) {
+      CallExpression callExpression1 = (CallExpression) argument.expression();
+      Symbol calleeSymbol = callExpression1.calleeSymbol();
+      if (!(calleeSymbol != null && "pytz.timezone".equals(calleeSymbol.fullyQualifiedName()))) {
+        return;
       }
+      context.addIssue(argument, MESSAGE);
+    } else if (argument.expression().is(Tree.Kind.NAME)) {
+      var lastAssignmentExpression = reachingDefinitionsAnalysis.valuesAtLocation((Name) argument.expression()).stream()
+        .filter(expression -> expression.is(Tree.Kind.CALL_EXPR))
+        .map(CallExpression.class::cast)
+        .findFirst();
+      var lastAssignmentSymbol = lastAssignmentExpression.map(CallExpression::calleeSymbol)
+        .filter(symbol -> "pytz.timezone".equals(symbol.fullyQualifiedName()));
+
+      if (lastAssignmentExpression.isEmpty()) {
+        return;
+      }
+      lastAssignmentSymbol.ifPresent(assignment -> context.addIssue(argument, MESSAGE).secondary(lastAssignmentExpression.get(), SECONDARY_MESSAGE));
```

Comment:
I find the logic here a bit imprecise.

`ReachingDefinitionsAnalysis#valuesAtLocation` can actually return multiple values as it is not path sensitive:
```
def foo(cond):
  if cond:
    my_tz = get_some_tz()
  else:
    my_tz = pytz.timezone(...)
  datetime.datetime(..., tzinfo=my_tz) # Reaching definitions are both get_some_tz() and pyt.timezone(...)
```

By calling `findFirst()` you'll keep the first one of those two definitions you encounter. In the best case, if the ordering was based on syntax, it would be a bit imprecise and could simply lead to some FPs/FNs depending on how the code is written.
It's actually worse, as `ReachingDefinitionsAnalysis#valuesAtLocation` returns a `Set` which means that the order of iteration is not guaranteed and the element you retrieve may be non-deterministic (which would in turn make whether we raise the issue or not non-deterministic).

Rather than looking for the first element we can find, I think we need to make a conscious choice: do all of the reaching definitions need to be noncompliant for us to report the issue? Or is it enough that one of the reaching definition is noncompliant? Depending on that choice, I think this snippet should be refactored to be a bit more intentional.

Also, some additional tests will be needed to document the behavior when the object has multiple definitions.